File: src/constraints.ts
import { SetRequired } from "type-fest";
import { generalizeType, resolveType } from "./analyze";
import { dumpNode } from "./ast";
import { getLastExpression } from "./lib/getLastExpression";
import {
  BinaryOperation,
  Block,
  BodyItem,
  ConstDeclaration,
  DotNotationCall,
  EnumCall,
  EnumCallType,
  EnumDeclaration,
  EnumPattern,
  Expression,
  FunctionCall,
  FunctionCallType,
  FunctionExpression,
  FunctionType,
  Identifier,
  IfElseExpression,
  MatchExpression,
  NativeType,
  NodeType,
  ObjectLiteral,
  ObjectPropLike,
  ObjectType,
  ParameterType,
  Pattern,
  PatternType,
  Program,
  PropertyTypeDefinition,
  Statement,
  TypeDeclaration,
  TypeExpression,
  TypeReference,
} from "./nodes";
import {
  ConstraintKind,
  ConstraintType,
  Scope,
  createTypeSymbol,
  createTypeVariable,
  dumpScope,
  findTypeSymbol,
  findValueSymbol,
  pushConstraint,
  renderTypeNode,
} from "./scope";
import { deepEquals } from "bun";

export function collectProgram(program: SetRequired<Program, "scope">) {
  program.body?.forEach((item) => collectBodyItem(item, program.scope));
  return program;
}

export function collectBodyItem(
  bodyItem: BodyItem,
  scope: Scope,
): ConstraintType {
  if (bodyItem.type === NodeType.ConstDeclaration) {
    return collectConstDeclaration(bodyItem, scope);
  } else if (bodyItem.type === NodeType.EnumDeclaration) {
    return collectEnumDeclaration(bodyItem, scope);
  } else if (bodyItem.type === NodeType.TypeDeclaration) {
    return collectTypeDeclaration(bodyItem, scope);
  } else {
    return collectStatement(bodyItem, scope);
  }
}

export function collectConstDeclaration(
  constDec: ConstDeclaration,
  scope: Scope,
): ConstraintType {
  if (constDec.name.type === NodeType.EnumDestructureBinding) {
    const name = constDec.name.unwrap[0].name;
    const existingTypeVar = scope.value[name]?.type;

    if (!existingTypeVar) {
      throw new Error(`Type variable for ${name} not found`);
    }

    const exprType = collectExpression(constDec.value, scope);

    // Add constraint between the existing type variable and the expression type
    // console.log("constdec", {
    //   existingTypeVar: dumpNode(existingTypeVar),
    //   exprType: dumpNode(exprType),
    //   constDec: dumpNode(constDec),
    // });

    const enumType = findTypeSymbol(constDec.name.enumName.name, scope)?.type;
    if (!enumType) {
      throw new Error(
        `Couldn't find an enum named ${constDec.name.enumName.name}`,
      );
    }
    const patternMatch = <PatternType>{
      type: NodeType.PatternType,
      typeVar: existingTypeVar,
      pattern: <EnumPattern>{
        type: NodeType.EnumPattern,
        enum: enumType,
        member: constDec.name.memberName,
      },
    };
    scope.constraints.push([patternMatch, exprType, scope]);
    return existingTypeVar;
  } else {
    const name =
      constDec.name.type === NodeType.Identifier
        ? constDec.name.name
        : constDec.name.identifiers[0].name; // Assuming first for simplification, handle more later

    const existingTypeVar = scope.value[name]?.type;

    if (!existingTypeVar) {
      throw new Error(`Type variable for ${name} not found`);
    }

    const exprType = collectExpression(constDec.value, scope);

    pushConstraint(scope, [existingTypeVar, exprType]);

    // special sauce to handle const-bound functions
    if (constDec.value.type === NodeType.FunctionExpression) {
      // @ts-ignore
      scope.type[existingTypeVar?.name].type = constDec.value.identifier!;
    }

    return existingTypeVar;
  }
}

export function collectTypeDeclaration(
  typeDec: TypeDeclaration,
  _scope: Scope,
) {
  const typeSymbol = findTypeSymbol(typeDec.identifier.name, _scope);
  if (!typeSymbol)
    throw new Error(
      `Could not find type symbol named ${typeDec.identifier.name} but it should have been creating by bindTypeDeclaration`,
    );
  // TODO: possible need to do some collection here?
  typeSymbol.type = typeDec.value;

  return <TypeReference>{
    type: NodeType.TypeReference,
    identifier: typeDec.identifier,
  };
}

export function collectStatement(
  statement: Statement,
  scope: Scope,
): TypeExpression {
  if (statement.type === NodeType.Block) {
    return collectBlock(statement);
  } else if (statement.type === NodeType.DebuggerStatement) {
    return { type: NodeType.NativeType, kind: "void" };
  } else {
    // Must be an expression
    return collectExpression(statement, scope);
  }
}

export function collectBlock(block: Block): TypeExpression {
  const blockScope = block.scope;

  if (!blockScope) {
    throw new Error("Block scope not found");
  }

  if (!block.body) return { type: NodeType.NativeType, kind: "void" };

  const constraints = block.body.map((stmt) => {
    return collectBodyItem(stmt, blockScope);
  });

  return constraints.at(-1) ?? { type: NodeType.NativeType, kind: "void" };
}

export function collectExpression(
  expr: Expression,
  scope: Scope,
): ConstraintType {
  switch (expr.type) {
    case NodeType.PrimitiveValue:
      return {
        type: NodeType.NativeType,
        kind: expr.kind,
      };

    case NodeType.BinaryOperation:
      return collectBinaryOperation(expr, scope);

    case NodeType.FunctionExpression:
      return collectFunctionDefinition(
        // @ts-ignore
        expr as SetRequired<FunctionExpression, "scope">,
      );

    case NodeType.FunctionCall:
      return collectFunctionCall(expr, scope);

    case NodeType.Identifier:
      return collectIdentifier(expr, scope);

    case NodeType.IfElseExpression:
      return collectIfElseExpression(expr, scope);

    case NodeType.DotNotationCall:
      return collectDotNotationCall(expr, scope);

    case NodeType.TemplateLiteral: {
      expr.spans
        .filter((span) => span.type === NodeType.TemplateSpan)
        .forEach((templateSpan) =>
          pushConstraint(scope, [
            collectExpression(templateSpan.expression, scope),
            <NativeType>{ type: NodeType.NativeType, kind: "string" },
          ]),
        );
      return { type: NodeType.NativeType, kind: "string" };
    }

    case NodeType.ObjectLiteral: {
      return collectObjectLiteral(expr, scope);
    }

    case NodeType.EnumCall: {
      return collectEnumCall(expr, scope);
    }

    case NodeType.MatchExpression: {
      return collectMatchExpression(expr, scope);
    }
  }

  throw new Error(`TODO: collectExpression with ${NodeType[expr.type]}`);
}

export function collectMatchExpression(
  matchExpr: MatchExpression,
  scope: Scope,
): TypeExpression {
  const subjectType = collectExpression(matchExpr.subject, scope);
  const firstClause = matchExpr.clauses[0];
  if (!firstClause) {
    throw new Error(`match expression has no clauses`);
  }

  const clauseTypes: [TypeExpression, TypeExpression][] = matchExpr.clauses.map(
    (clause) => [
      collectPattern(clause.pattern, clause.scope!),
      collectStatement(clause.body, clause.scope!),
    ],
  );

  const [firstClauseType, ...otherClauses] = clauseTypes;
  const [firstPatternType, firstBodyType] = firstClauseType;

  console.log("clauseTypes", dumpNode(clauseTypes));

  scope.constraints.push([subjectType, firstPatternType, scope]);

  otherClauses.forEach(([patternType, bodyType]) => {
    scope.constraints.push([
      patternType,
      subjectType,
      scope,
      ConstraintKind.Subset,
    ]);
    scope.constraints.push([firstBodyType, bodyType, scope]);
  });

  return firstBodyType;
}

export function collectPattern(pattern: Pattern, scope: Scope): TypeExpression {
  return collectExpression(pattern as Expression, scope);
}

export function collectEnumDeclaration(
  enumDec: EnumDeclaration,
  scope: Scope,
): TypeExpression {
  const enumTypeSymbol = findTypeSymbol(enumDec.identifier.name, scope);
  if (!enumTypeSymbol)
    throw new Error(`Missing EnumType for ${enumDec.identifier.name}`);
  return enumTypeSymbol.type;
}

export function collectEnumCall(
  enumCall: EnumCall,
  scope: Scope,
): TypeExpression {
  if (enumCall.expression.type !== NodeType.DotNotationCall) {
    throw new Error(`Odd way to call an enum??`);
  }

  const leftType = collectExpression(enumCall.expression.left, scope);
  if (!leftType || leftType.type !== NodeType.EnumType) {
    console.log(dumpNode(leftType));
    throw new Error(`Left side did not resolve to an enum type`);
  }

  const enumMember = leftType.members.find(
    (member) => member.identifier.name === enumCall.expression.right.name,
  );

  if (!enumMember) {
    throw new Error(
      `Enum ${leftType.identifier.name} does not have a variant called ${enumCall.expression.right.name}`,
    );
  }
  const argTypes = enumCall.arguments.map((arg) =>
    collectExpression(arg, scope),
  );

  const enumCallType = <EnumCallType>{
    type: NodeType.EnumCallType,
    enum: leftType,
    member: enumMember,
    arguments: argTypes,
  };

  return enumCallType;
}

export function collectObjectLiteral(
  object: ObjectLiteral,
  scope: Scope,
): TypeExpression {
  const objectType: ObjectType = {
    type: NodeType.ObjectType,
    definitions: object.properties.map((prop) => {
      return <PropertyTypeDefinition>{
        type: NodeType.PropertyTypeDefinition,
        name: prop.name,
        value: collectExpression(prop.value, scope),
      };
    }),
  };
  return objectType;
}

export function collectDotNotationCall(
  dotCall: DotNotationCall,
  scope: Scope,
): TypeExpression {
  const referencedType = collectExpression(dotCall.left, scope);

  if (referencedType.type === NodeType.EnumType) {
    const referencedMember = referencedType.members.find((memberType) => {
      return deepEquals(memberType.identifier, dotCall.right);
    });
    if (!referencedMember)
      throw new Error(`Couldn't find enum member called ${dotCall.right.name}`);

    return <EnumCallType>{
      type: NodeType.EnumCallType,
      enum: referencedType,
      member: referencedMember,
      arguments: [],
    };
  } else if (referencedType.type === NodeType.InferenceRequired) {
    if (dotCall.right.type !== NodeType.Identifier) {
      // this should be unreachable due to the grammar
      throw new Error(`Dot call expressions must be static identifiers`);
    }
    const propTypeVar = createTypeVariable(scope);
    // unknown type, so we need to infer
    const objectType: ObjectType = {
      type: NodeType.ObjectType,
      identifier: createTypeVariable(scope),
      definitions: [
        <PropertyTypeDefinition>{
          type: NodeType.PropertyTypeDefinition,
          name: dotCall.right,
          value: propTypeVar,
        },
      ],
    };
    pushConstraint(scope, [referencedType, objectType]);
    return propTypeVar;
  } else if (referencedType.type === NodeType.ObjectType) {
    const referencedPropDefinition = referencedType.definitions.find((def) =>
      deepEquals(def.name, dotCall.right),
    );
    console.log("ref", dumpNode(referencedType));
    console.log("dotCall", dumpNode(dotCall));
    if (!referencedPropDefinition) {
      throw new Error(`Could not find a field called ${dotCall.right.name}`);
    }
    console.log("refDef", dumpNode(referencedPropDefinition));
    return referencedPropDefinition.value;
  } else {
    throw new Error("unhandled dot call notation");
  }
  throw new Error("**");
}

export function collectIfElseExpression(
  expr: IfElseExpression,
  scope: Scope,
): TypeExpression {
  const conditionType = collectExpression(expr.condition, scope);
  pushConstraint(scope, [
    conditionType,
    <NativeType>{ type: NodeType.NativeType, kind: "boolean" },
  ]);

  const trueBranch = collectStatement(expr.trueBlock, scope);
  const falseBranch = collectStatement(expr.falseBlock, scope);

  pushConstraint(scope, [trueBranch, falseBranch]);

  return generalizeType(trueBranch);
}

export function collectBinaryOperation(
  expr: BinaryOperation,
  scope: Scope,
): TypeExpression {
  const leftType = collectExpression(expr.left, scope);
  const rightType = collectExpression(expr.right, scope);

  // Add constraints based on the operator.
  switch (expr.op) {
    case "addition":
    case "subtraction":
    case "multiply":
    case "divide":
    case "exponentiation":
      pushConstraint(scope, [
        leftType,
        { type: NodeType.NativeType, kind: "number" },
      ]);
      pushConstraint(scope, [
        rightType,
        { type: NodeType.NativeType, kind: "number" },
      ]);
      return { type: NodeType.NativeType, kind: "number" };

    case "greaterThan":
    case "greaterOrEqual":
    case "lessThan":
    case "lessOrEqual":
      pushConstraint(scope, [
        leftType,
        { type: NodeType.NativeType, kind: "number" },
      ]);
      pushConstraint(scope, [
        rightType,
        { type: NodeType.NativeType, kind: "number" },
      ]);
      return { type: NodeType.NativeType, kind: "boolean" };

    case "logicalOr":
    case "logicalAnd":
      // These must be booleans
      pushConstraint(scope, [
        leftType,
        { type: NodeType.NativeType, kind: "boolean" },
      ]);
      pushConstraint(scope, [
        rightType,
        { type: NodeType.NativeType, kind: "boolean" },
      ]);
      return { type: NodeType.NativeType, kind: "boolean" };

    case "equal":
    case "notEqual":
      pushConstraint(scope, [leftType, rightType]);
      return { type: NodeType.NativeType, kind: "boolean" };
  }
  throw new Error(`Unknown binary operation ${expr.op}`);
}

export function collectFunctionCall(
  fnCall: FunctionCall,
  scope: Scope,
): TypeExpression {
  const referencedValue = collectExpression(fnCall.expression, scope);
  const resolvedValue = resolveType(referencedValue, scope);

  const isResolvedFunction = resolvedValue.type === NodeType.FunctionType;
  const isUnresolvedReference =
    resolvedValue.type === NodeType.InferenceRequired ||
    resolvedValue.type === NodeType.Identifier;

  const fnCallType: FunctionCallType = {
    type: NodeType.FunctionCallType,
    returnType: createTypeVariable(scope),
    arguments: fnCall.arguments.map((arg) => collectExpression(arg, scope)),
    callee: referencedValue,
  };

  if (isUnresolvedReference) {
    // if we can't resolve the FunctionType being called,
    // create a constraint that resolvedValue must be a FunctionType
    const fnType: FunctionType = {
      type: NodeType.FunctionType,
      returnType: {
        type: NodeType.TypeAnnotation,
        expression: fnCallType.returnType,
      },
      identifier:
        resolvedValue.type === NodeType.Identifier ||
        resolvedValue.type === NodeType.InferenceRequired
          ? resolvedValue
          : undefined,
      parameters: fnCall.arguments.map((arg: Expression): ParameterType => {
        return {
          type: NodeType.ParameterType,
          typeAnnotation: {
            type: NodeType.TypeAnnotation,
            expression: collectExpression(arg, scope),
          },
          isSpread: false, // TODO: spread parameter support
        };
      }),
    };

    pushConstraint(scope, [resolvedValue, fnType]);
  } else if (isResolvedFunction) {
    pushConstraint(scope, [resolvedValue, fnCallType]);
  } else {
    throw new Error(
      `Cannot call ${renderTypeNode(resolvedValue)} as it is not a function`,
    );
  }

  return fnCallType.returnType;
}

export function collectFunctionDefinition(
  fnExpression: SetRequired<FunctionExpression, "scope" | "identifier">,
): TypeExpression {
  const fnScope = fnExpression.scope;
  const fnTypeSymbol = findTypeSymbol(fnExpression.identifier.name, fnScope);

  if (!fnTypeSymbol) {
    throw new Error("Function type not found in type symbol table");
  }

  if (fnTypeSymbol.type.type !== NodeType.FunctionType) {
    console.log(
      "FnExpression: %o\nSymbol: %o\n",
      dumpNode(fnExpression),
      dumpNode(fnTypeSymbol.type),
    );
    throw new Error(`Related symbol isn't a function type`);
  }

  const returnType = fnTypeSymbol.type.returnType.expression;

  // Collect constraints for the function body
  collectStatement(fnExpression.body, fnExpression.scope);

  const [returnExpression, returnScope] = getLastExpression(
    fnExpression.body,
    fnScope,
  );
  const returnExpressionType = returnExpression
    ? collectExpression(returnExpression, returnScope)
    : ({ type: NodeType.NativeType, kind: "void" } as NativeType);

  if (returnExpressionType) {
    if (!returnType) throw new Error("no return type");
    console.log("FD", {
      [renderTypeNode(returnType)]: renderTypeNode(returnExpressionType),
    });
    pushConstraint(fnScope, [returnExpressionType, returnType]);
  }

  const fnType = findTypeSymbol(
    fnExpression.identifier.name,
    fnExpression.scope,
  );
  if (!fnType)
    throw new Error(
      `Could not find function named ${fnExpression.identifier.name}`,
    );

  return fnType.type;
}

export function collectIdentifier(
  identifier: Identifier,
  scope: Scope,
): ConstraintType {
  const identifierType = findValueSymbol(identifier.name, scope);
  if (!identifierType) {
    throw new Error(`Couldn't find type for ${identifier.name}`);
  }

  if (!identifierType.type)
    throw new Error(`No type information for identifier ${identifier.name}`);

  return identifierType.type!;
}

File: src/constraints.rs
use core::fmt;
use core::panic;

use crate::{ast::*, scope::ScopeTree};

#[derive(Debug, Clone, PartialEq)]
pub enum ConstraintKind {
    Equality,
    Subset,
    PatternMatch,
}

impl fmt::Display for ConstraintKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ConstraintKind::Equality => write!(f, "="),
            ConstraintKind::Subset => write!(f, "><"),
            ConstraintKind::PatternMatch => write!(f, "matches"),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Constraint {
    pub lhs: TypeExpr,
    pub rhs: TypeExpr,
    pub kind: ConstraintKind,
    pub scope_index: usize,
}

#[derive(Debug)]
pub struct ConstraintCollector<'a> {
    scope_tree: &'a mut ScopeTree,
    pub constraints: Vec<Constraint>,
}

impl<'a> ConstraintCollector<'a> {
    pub fn new(scope_tree: &'a mut ScopeTree) -> Self {
        ConstraintCollector {
            scope_tree,
            constraints: Vec::new(),
        }
    }

    pub fn push_constraint(&mut self, constraint: Constraint) {
        self.constraints.push(constraint);
    }

    pub fn collect_program(&mut self, program: Program) -> Program {
        let _: Vec<TypeExpr> = program
            .clone()
            .statements
            .iter()
            .map(|expr| match expr {
                TopStatement::ConstDec(const_dec) => {
                    self.collect_const_dec(const_dec.clone(), program.clone().scope.unwrap())
                }
                TopStatement::TypeDec(_) => todo!(),
                TopStatement::Expr(expr) => {
                    self.collect_expr(expr.clone(), program.clone().scope.unwrap())
                }
                TopStatement::EnumDec(_) => todo!(),
                TopStatement::ExternDec(extern_package) => {
                    self.collect_extern_dec(extern_package.clone(), program.clone().scope.unwrap())
                }
            })
            .collect();
        program
    }

    fn collect_statement(&mut self, statement: BlockStatement, parent_scope: usize) -> TypeExpr {
        match statement {
            BlockStatement::ConstDec(const_dec) => {
                self.collect_const_dec(const_dec.clone(), parent_scope)
            }
            BlockStatement::Return(expr) => self.collect_expr(expr, parent_scope),
            BlockStatement::Expr(expr) => self.collect_expr(expr.clone(), parent_scope),
        }
    }

    fn collect_extern_dec(&mut self, extern_dec: ExternPackage, parent_scope: usize) -> TypeExpr {
        // extern dec already has a fully fleshed out type in the scope tree
        // this is done during scope binding, lets just pull that out
        let type_symbol = self
            .scope_tree
            .find_type_symbol(
                parent_scope,
                TypeIdentifier {
                    name: vec![extern_dec.package_name],
                },
            )
            .expect("extern dec should already have type symbol from binding phase");

        type_symbol.type_expr
    }

    fn collect_const_dec(&mut self, const_dec: ConstDec, parent_scope: usize) -> TypeExpr {
        let name = const_dec.clone().identifier.name;
        let value_symbol = self
            .scope_tree
            .find_value_symbol(parent_scope, &name)
            .unwrap();
        let const_type = value_symbol.type_expr;
        let expr_type = self.collect_expr(*const_dec.clone().value, parent_scope);

        self.push_constraint(Constraint {
            lhs: const_type.clone(),
            rhs: expr_type,
            kind: ConstraintKind::Equality,
            scope_index: parent_scope,
        });

        const_type
    }

    fn collect_expr(&mut self, expr: Expr, parent_scope: usize) -> TypeExpr {
        match expr {
            Expr::Number(_) => TypeExpr::Number,
            Expr::String(_) => TypeExpr::String,
            Expr::Boolean(_) => TypeExpr::Boolean,
            Expr::FunctionDefinition {
                parameters: _,
                return_type: _return_type,
                body,
                scope: Some(fn_scope),
                identifier: Some(identifier),
            } => {
                println!("fn_name {:#?}", identifier.clone());
                let as_type_iden = TypeIdentifier {
                    name: vec![identifier.clone().name],
                };
                let fn_type_symbol = self
                    .scope_tree
                    .find_type_symbol(fn_scope, as_type_iden)
                    .expect("Error: expected fn_def type symbol");

                let return_type = match fn_type_symbol.clone().type_expr {
                    TypeExpr::FunctionDefinition { return_type, .. } => *return_type,
                    _ => panic!(
                        "Fn by name of {} is not a Function type in the symbol table",
                        identifier.name
                    ),
                };

                let body_returns = self.collect_expr(*body, fn_scope);
                self.push_constraint(Constraint {
                    lhs: return_type,
                    rhs: body_returns,
                    kind: ConstraintKind::Equality,
                    scope_index: fn_scope,
                });

                // need to ensure identifier is set before here?
                // bind_const_dec sets it, but what bout anonymous name?
                fn_type_symbol.type_expr
            }
            Expr::ValueReference(mixed_identifier) => {
                let iden_name = match mixed_identifier {
                    MixedIdentifier::Identifier(identifier) => identifier.clone().name,
                    MixedIdentifier::TypeIdentifier(type_identifier) => type_identifier
                        .clone()
                        .name
                        .first()
                        .expect("at least one name")
                        .to_string(),
                };
                println!(
                    "looking up value ref {} in scope {}",
                    iden_name, parent_scope
                );
                let value_symbol = self.scope_tree.find_value_symbol(parent_scope, &iden_name);
                if value_symbol.is_none() {
                    panic!("Could not find {:?} in socpe", iden_name);
                }
                value_symbol.unwrap().type_expr
            }
            // Expr::TypeDec(type_dec) => {
            //     let type_symbol = self
            //         .scope_tree
            //         .find_type_symbol(parent_scope, type_dec.identifier.name);
            //
            //     type_symbol.unwrap().type_expr
            // }
            Expr::Record(_, _) => todo!(),
            Expr::Array(array_type, exprs) => {
                for expr in exprs {
                    let expr_type = self.collect_expr(expr.clone(), parent_scope);
                    self.push_constraint(Constraint {
                        lhs: array_type.clone(),
                        rhs: expr_type,
                        kind: ConstraintKind::Equality,
                        scope_index: parent_scope,
                    })
                }
                array_type
            }
            Expr::BlockExpression(statements, scope_index) => {
                let block_scope = scope_index.unwrap();
                let returned_exprs: Vec<TypeExpr> = statements
                    .iter()
                    .filter_map(|statement| match statement {
                        BlockStatement::Return(expr) => {
                            Some(self.collect_expr(expr.clone(), block_scope))
                        }
                        BlockStatement::ConstDec(const_dec) => {
                            // this is yuck? fix mutation in filter_map
                            let _ = self.collect_const_dec(const_dec.clone(), block_scope);
                            None
                        }
                        BlockStatement::Expr(expr) => {
                            // this is yuck? fix mutation in filter_map
                            let _ = self.collect_expr(expr.clone(), block_scope);
                            None
                        }
                    })
                    .collect();
                let last_return = returned_exprs.last().unwrap_or(&TypeExpr::Void);
                for returned_expr in returned_exprs.clone() {
                    if returned_expr != *last_return {
                        self.push_constraint(Constraint {
                            lhs: last_return.clone(),
                            rhs: returned_expr,
                            kind: ConstraintKind::Equality,
                            scope_index: block_scope,
                        })
                    }
                }
                last_return.clone()
            }
            Expr::Void => TypeExpr::Void,
            Expr::Binary(left, _op, right) => {
                let left_type = self.collect_expr(*left, parent_scope);
                let right_type = self.collect_expr(*right, parent_scope);
                self.push_constraint(Constraint {
                    lhs: left_type.clone(),
                    rhs: right_type.clone(),
                    kind: ConstraintKind::Equality,
                    scope_index: parent_scope,
                });
                let push_numbers_constraint = |return_type| {
                    self.push_constraint(Constraint {
                        lhs: TypeExpr::Number,
                        rhs: left_type,
                        kind: ConstraintKind::Equality,
                        scope_index: parent_scope,
                    });
                    self.push_constraint(Constraint {
                        lhs: TypeExpr::Number,
                        rhs: right_type,
                        kind: ConstraintKind::Equality,
                        scope_index: parent_scope,
                    });
                    return_type
                };
                match _op {
                    BinaryOp::Add => push_numbers_constraint(TypeExpr::Number),
                    BinaryOp::Subtract => push_numbers_constraint(TypeExpr::Number),
                    BinaryOp::Multiply => push_numbers_constraint(TypeExpr::Number),
                    BinaryOp::Divide => push_numbers_constraint(TypeExpr::Number),
                    BinaryOp::Equal => TypeExpr::Boolean,
                    BinaryOp::NotEqual => TypeExpr::Boolean,
                    BinaryOp::GreaterThan => push_numbers_constraint(TypeExpr::Boolean),
                    BinaryOp::GreaterOrEqual => push_numbers_constraint(TypeExpr::Boolean),
                    BinaryOp::LessThan => push_numbers_constraint(TypeExpr::Boolean),
                    BinaryOp::LessOrEqual => push_numbers_constraint(TypeExpr::Boolean),
                }
            }
            Expr::DotCall(callee, member_identifier) => {
                let callee_type = self.collect_expr(*callee, parent_scope);
                let resolved_callee_type = self
                    .scope_tree
                    .resolve_type(callee_type.clone(), parent_scope);

                match resolved_callee_type {
                    TypeExpr::ExternPackage {
                        package_name,
                        members,
                    } => {
                        let member_search = members.iter().find(|&member| match member {
                            ExternMember::Function { local_name, .. } => {
                                *local_name == member_identifier
                            }
                            ExternMember::Variable { local_name, .. } => {
                                *local_name == member_identifier
                            }
                        });

                        match member_search {
                            Some(member) => match &member {
                                ExternMember::Function {
                                    local_name,
                                    external_name: _,
                                    parameters,
                                    return_type,
                                } => {
                                    let type_identifier = TypeIdentifier {
                                        name: vec![package_name, local_name.clone().name],
                                    };
                                    let params_types = parameters
                                        .iter()
                                        .map(|p| p.type_expr.clone().unwrap())
                                        .collect();

                                    TypeExpr::FunctionDefinition {
                                        type_identifier,
                                        parameters: params_types,
                                        return_type: Box::new(return_type.clone()),
                                    }
                                }
                                ExternMember::Variable {
                                    local_name,
                                    external_name,
                                    value_type,
                                } => todo!(),
                            },
                            None => {
                                panic!("Cannot call {} on {}", member_identifier.name, package_name)
                            }
                        }
                    }
                    TypeExpr::ImportRef(name, module_indexes) => {
                        println!("name: {}", name);
                        println!("module_indexes: {:#?}", module_indexes);
                        self.scope_tree
                            .resolve_import_member_type(name.clone(), member_identifier.clone())
                            .unwrap_or_else(|| {
                                panic!("can resolve {}.{}", name, member_identifier.name)
                            })
                    }
                    _ => {
                        panic!(
                            "Unhandled dotcall type {:#?} /endunhandled",
                            resolved_callee_type
                        );
                    }
                }
            }
            Expr::FunctionCall {
                callee,
                args,
                generic_args: _,
            } => {
                let callee_type = self.collect_expr(*callee, parent_scope);
                let resolved_type = self
                    .scope_tree
                    .resolve_type(callee_type.clone(), parent_scope);

                let already_resolves_to_fn = matches!(
                    resolved_type,
                    TypeExpr::FunctionDefinition { .. } | TypeExpr::FunctionCall { .. }
                );

                let return_type = self.scope_tree.create_type_var(parent_scope);
                let fn_call_type = TypeExpr::FunctionCall {
                    args: args
                        .iter()
                        .map(|arg| self.collect_expr(arg.clone(), parent_scope))
                        .collect(),
                    return_type: Box::new(return_type.clone()),
                    callee: Box::new(callee_type.clone()),
                };

                if already_resolves_to_fn {
                    self.push_constraint(Constraint {
                        lhs: fn_call_type,
                        rhs: resolved_type,
                        kind: ConstraintKind::Equality,
                        scope_index: parent_scope,
                    });
                } else {
                    let identifier = match resolved_type.clone() {
                        TypeExpr::TypeRef(type_identifier)  => type_identifier.clone(),
                        TypeExpr::InferenceRequired(Some(type_identifier)) => type_identifier.clone(),
                        _ => panic!("Expected fn call resolved type to be a TypeIdentifier or InferferenceRequired(Some(TypeIdentifier)). Got {:#?}", resolved_type.clone()),
                    };
                    // existing fn expression is resolving to something like fn1
                    // lets infer a function def based on the call type
                    let fn_def_type = TypeExpr::FunctionDefinition {
                        type_identifier: identifier,
                        parameters: args
                            .iter()
                            .map(|a| self.collect_expr(a.clone(), parent_scope))
                            .collect(),
                        return_type: Box::new(return_type.clone()),
                    };

                    self.push_constraint(Constraint {
                        lhs: resolved_type,
                        rhs: fn_def_type,
                        kind: ConstraintKind::Equality,
                        scope_index: parent_scope,
                    })
                }

                return_type
            }
            Expr::Match(_, _) => todo!(),
            Expr::IfElse(condition, true_branch, false_branch) => {
                let condition_type = self.collect_expr(*condition, parent_scope);
                let true_branch_type = self.collect_expr(*true_branch, parent_scope);
                let false_branch_type = self.collect_expr(*false_branch, parent_scope);

                self.push_constraint(Constraint {
                    lhs: condition_type,
                    rhs: TypeExpr::Boolean,
                    kind: ConstraintKind::Equality,
                    scope_index: parent_scope,
                });
                self.push_constraint(Constraint {
                    lhs: true_branch_type.clone(),
                    rhs: false_branch_type,
                    kind: ConstraintKind::Equality,
                    scope_index: parent_scope,
                });
                true_branch_type
            }
            Expr::FunctionDefinition { .. } => panic!("Fn def has something missing"),
        }
    }
}
